### PR TITLE
DEV | Space Members  |  Join Space  |  Leave Space  |  View Space Members  |  Post Edit

### PR DESCRIPTION
**Pull request by sehmuzc**
### Description

- 1-Post Editor HTML page created to fix the problem at post edits.
- 2-Home page post list updated to also gather the posts from the spaces that the user is owner, member or moderator.
- 3-SpaceForm has been deleted from forms.py in Space app as refactoring. 
- 4-Members and Moderators of a space can be empty. This may be required when making changes from admin panel. 
- 5-Join and Leave Space functionality added.
- 6-View members of a space functionality added to see the moderators and members in a given space. 
- 7-ContactAdmin has been added to Admin part.

### CODE REVIEW

File: posts/templates/post_editor.html [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                }
+            }
+        });
+    });
+});
+</script>
+
+
+
+{% endblock %}
```
Comment (ceydaduzgec) [inline posts/templates/post_editor.html:103]: Could you add a new empty line to 104 please?

File: posts/urls.py [hunk 2]
```diff
@@ -23,7 +23,7 @@
     path("<int:labels>/", post_list, name="post_list"),
     path("post/new/", post_new, name="post_new"),
     path("post/<int:pk>/share/", post_share, name="post_share"),
-    path("post/<int:pk>/edit/", post_edit, name="post_edit"),
+    path("post/<int:pk>/editor/", post_edit, name="post_edit"),
```
Comment (ceydaduzgec) [inline posts/urls.py:26]: Why did you change edit to editor?

File: posts/views.py [hunk 3]
```diff
@@ -30,15 +30,17 @@ def post_list(request, tag_slug=None):
...
+    spaces = Space.objects.filter(
+        Q(owner=request.user) | Q(members=request.user) | Q(granted_members=request.user) | Q(moderators=request.user)
     )
 
+    # Getting posts of the user and their friends, and posts from the spaces
+    posts = Post.objects.filter(
+        Q(author__in=friendsofUser) | Q(author=request.user) | Q(spaces__in=spaces),
+        published_date__lte=timezone.now()
+    ).annotate(total_comments=Count("comments")).order_by("-published_date")
+
```
Comment (ceydaduzgec) [inline posts/views.py:43]: I am not gonna test this part now, we can create a bug issue later if it doesn't work.

File: posts/views.py [hunk 4]
```diff
@@ -48,6 +50,7 @@ def post_list(request, tag_slug=None):
     return render(request, "post_list.html", {"posts": posts, "tag": tag})
 
 
+
```
Comment (ceydaduzgec) [inline posts/views.py:53]: Can you delete this extra third line? Unrelated codes parts only need to have 2 empty lines between them.

File: posts/views.py [hunk 5]
```diff
@@ -98,11 +101,25 @@ def post_edit(request, pk):
...
+            post.image = request.FILES.get('image')
+            post.tags.clear()  # clear existing tags
+            tags = form.cleaned_data['tags']
+            if isinstance(tags, list):
+                tags = ','.join(tags)
+            tags_list = [tag.strip() for tag in tags.split(',')]
+            for tag in tags_list:
+                if tag:
+                    t, created = Tag.objects.get_or_create(name=tag)
+                    post.tags.add(t)
```
Comment (ceydaduzgec) [inline posts/views.py:113]: This part is nicely done thanks :)

File: posts/views.py [hunk 6]
```diff
@@ -98,11 +101,25 @@ def post_edit(request, pk):
...
             post.save()
             return redirect("post_detail", pk=post.pk)
     else:
         form = PostForm(instance=post)
-    return render(request, "post_edit.html", {"form": form})
+    return render(request, "post_editor.html", {"form": form, "post": post})
+
+
+
+
```
Comment (ceydaduzgec) [inline posts/views.py:122]: Can you also delete these extra lines and make it only 2 lines?

File: spaces/forms.py [hunk 7]
```diff
@@ -16,20 +16,12 @@ class Meta:
...
-    class Meta:
-        model = Post
-        fields = ("title", "link", "text")
-
-
 POSTING_PERMISSION_CHOICES = (
     ("all", "Any member can post"),
     ("granted", "Only granted members can post"),
     ("moderators", "Only moderators can post"),
 )
```
Comment (ceydaduzgec) [inline spaces/forms.py:24]: Can you add one line to make it 2 again?

File: spaces/models.py [hunk 8]
```diff
@@ -15,8 +15,8 @@ class Space(models.Model):
     owner = models.ForeignKey(User, on_delete=models.CASCADE, related_name="owned_spaces")
     name = models.CharField(max_length=200)
     description = models.TextField()
-    members = models.ManyToManyField(User, related_name="spaces")
-    moderators = models.ManyToManyField(User, related_name="moderated_spaces")
+    members = models.ManyToManyField(User, related_name="spaces", blank=True)
+    moderators = models.ManyToManyField(User, related_name="moderated_spaces", blank=True)
```
Comment (ceydaduzgec) [inline spaces/models.py:19]: Do we really want a space without any members and moderators? What is the purpose of the space then?

File: spaces/templates/space_detail.html [hunk 9]
```diff
@@ -3,6 +3,17 @@
 {% block content %}
   <h1>{{ space.name }}</h1>
   <p>{{ space.description }}</p>
+  {% if not request.user in space.members.all and not request.user in space.granted_members.all and not request.user in space.moderators.all %}
```
Comment (ceydaduzgec) [inline spaces/templates/space_detail.html:6]: This might create extra queries. Can you open an issue to investigate this? You can assign it to me.

File: spaces/templates/space_members.html [hunk 10]
```diff
@@ -0,0 +1,23 @@
+{% extends "base.html" %}
+
+{% block content %}
+  <h1>Members of {{ space.name }}</h1>
+
+  <h2>Moderators</h2>
+  <ul>
+    {% for moderator in moderators %}
+      <li>{{ moderator.username }}</li>
+    {% empty %}
```
Comment (ceydaduzgec) [inline spaces/templates/space_members.html:10]: Didn't know this, nice.

File: spaces/views.py [hunk 11]
```diff
@@ -5,14 +5,26 @@
...
+from spaces.forms import SpaceCreationForm
 from spaces.models import Space
 
 from .forms import SpacePolicyForm
 
 User = get_user_model()
 
+def join_space(request, pk):
+    space = get_object_or_404(Space, pk=pk)
+    space.members.add(request.user)
```
Comment (ceydaduzgec) [inline spaces/views.py:18]: Maybe we should check if user can add space twice. Did you check it? You can create issue for later if you want to.

File: spaces/views.py [hunk 12]
```diff
@@ -5,14 +5,26 @@
...
+    return redirect("space_detail", pk=pk)
+
+def leave_space(request, pk):
+    space = get_object_or_404(Space, pk=pk)
+    if request.user in space.members.all() or request.user in space.granted_members.all():
```
Comment (ceydaduzgec) [inline spaces/views.py:23]: We can create a mixin for this condition if it will be used other views.

Comment (ceydaduzgec) [reply spaces/views.py:23]: Also these should check the request and allow this only if it is a post request. Currently it allows get requests too I think.

File: users/admin.py [hunk 13]
```diff
@@ -1,8 +1,12 @@
...
-
+from .models import Contact
 
 @admin.register(Profile)
 class ProfileAdmin(admin.ModelAdmin):
     list_display = ["user", "date_of_birth", "photo"]
+
+@admin.register(Contact)
+class ContactAdmin(admin.ModelAdmin):
+    list_display = ('user_from', 'user_to', 'created')
```
Comment (ceydaduzgec) [inline users/admin.py:12]: Can you add a newline at the end please?

**Review by ceydaduzgec — DISMISSED** [review]
I left couple of comments. I think if you run linters and then review my comments it would be better since half of them are for syntax corrections. Thanks for your amazing work ✌️

**Review by eralp85 — APPROVED** [review]
Good work!

File: posts/templates/post_editor.html (hunk 1 above)
Comment (sehmuzc) [reply posts/templates/post_editor.html:103]: I will add it later.

File: posts/urls.py (hunk 2 above)
Comment (sehmuzc) [reply posts/urls.py:26]: Postedit was used for post creation. It was confusing so I wanted to have a different page to have the same functionality.

File: posts/views.py (hunk 4 above)
Comment (sehmuzc) [reply posts/views.py:53]: Will do it later.

File: posts/views.py (hunk 3 above)
Comment (sehmuzc) [reply posts/views.py:43]: You may join a space and list the posts after I merge the code.

File: spaces/templates/space_detail.html (hunk 9 above)
Comment (sehmuzc) [reply spaces/templates/space_detail.html:6]: Will open the issue later.

File: spaces/views.py (hunk 11 above)
Comment (sehmuzc) [reply spaces/views.py:18]: If user add spaces two times then two separate space will be opened.

File: users/admin.py (hunk 13 above)
Comment (sehmuzc) [reply users/admin.py:12]: Will update later.

File: spaces/templates/space_members.html (hunk 10 above)
Comment (sehmuzc) [reply spaces/templates/space_members.html:10]: Thanks.

File: spaces/views.py (hunk 12 above)
Comment (sehmuzc) [reply spaces/views.py:23]: You may update with the mixin, I am not familiar enough to create a new mixin, I just used the loginrequired mixins earlier. I was planning to check the space membership when viewing different steps.

File: spaces/models.py (hunk 8 above)
Comment (sehmuzc) [reply spaces/models.py:19]: Ceyda, I updated this part because I needed to create spaces without moderators to test different functionalities.

Comment (sehmuzc) [reply spaces/models.py:19]: Also the space module has members and granted members. If all the users are granted rights then the members can be blank. 
I plan to change the member to a granted member if the moderator gives the right to an ordinary member. The members can be empty.
